### PR TITLE
#652 - Probabilities from string matching recommender are not useful for active learning

### PR DESCRIPTION
**Pull request by reckart**
- Report the relative frequency of the label for a given word as the score
- Honor the maxRecommendations setting

### CODE REVIEW

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 1]
```diff
@@ -462,6 +504,23 @@ public void put(String aLabel)
...
         
+        public List<LabelCount> getBest(int aN)
+        {
+            List<LabelCount> best = new ArrayList<>();
+            for (int i = 0; i < labels.length; i++) {
+                best.add(new LabelCount(labels[i], counts[i]));
+            }
+            
+            if (labels.length > 1) {
+                System.out.println(".");
```
Comment (Rentier) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Remove this pls

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java [hunk 2]
```diff
@@ -126,8 +130,12 @@ public void predict(RecommenderContext aContext, CAS aCas) throws Recommendation
                     
                     // Need to check that the match actually ends at a token boundary!
                     if (tokens.stream().filter(t -> t.getEnd() == end).findAny().isPresent()) {
-                        spans.add(new Span(begin, end, text.substring(begin, end),
-                                node.value.getBestLabel()));
+                        int total = IntStream.of(node.value.counts).sum();
+                        for (LabelCount lc : node.value.getBest(maxRecommendations)) {
+                            double score = (double) lc.count / (double) total;
```
Comment (Rentier) [inline inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: I would describe the score you compute here.

Comment (reckart) [reply inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Done - although not here. I factored to code out into the LabelStats class and describe it there.

File: inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java (hunk 1 above)
Comment (reckart) [reply inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java]: Whoops :)

**Comment by reckart** [comment]
> I do not get recommendations from the string matcher anymore.

Hm - the unit tests work. I'll have a look.

**Comment by reckart** [comment]
> I do not get recommendations from the string matcher anymore.

The reason is probably that your recommender is configured to return no recommendations, i.e. "maxRecommendations = 0". Are you testing on an old existing project? Normally recommendation liquibase changeset "20181007-1" should have set the maxRecommendations value to 3. Or are you testing using a new project/recommender? The UI should not permit creating a recommender with maxRecommendations < 1 and > 100.

I'm fixing the `ExternalRecommenderIntegrationTest` to properly set the maxRecommendations field on its test recommender.

**Comment by reckart** [comment]
@Rentier Fix for recommendations not being generated in #669. Any remaining issues here to be addressed or good for approval?

**Comment by Rentier** [comment]
> > I do not get recommendations from the string matcher anymore.
> 
> The reason is probably that your recommender is configured to return no recommendations, i.e. "maxRecommendations = 0". Are you testing on an old existing project? Normally recommendation liquibase changeset "20181007-1" should have set the maxRecommendations value to 3. Or are you testing using a new project/recommender? The UI should not permit creating a recommender with maxRecommendations < 1 and > 100.
> 
> I'm fixing the `ExternalRecommenderIntegrationTest` to properly set the maxRecommendations field on its test recommender.

I tried both an old and a newly created project. The sidebar said 3 suggestions.

**Comment by Rentier** [comment]
If the recommendations work then this is done.

**Pull request merged**